### PR TITLE
Implement max duration of HTTP ConnectionPools

### LINKED ISSUE #5799: Allow specifying the duration an object can stay in a pool
**Jetty version**
9.4.x

**Description**
Currently, the `Pool` class keeps the objects it contains for an indefinite duration. While the number of times a pooled object can be used before it has to be closed can be configured which allows limiting the maximum time an object can stay in its pool, this can seriously shorten their useful lifespan if they are heavily used.

Introducing a new configurable property that specifies exactly how long an object can be pooled before it needs to be cleaned up would guarantee that pooled objects have both a maximum and a minimum lifespan. This would be useful for instance to help failing over HTTP connections in a predictable time to another server after the destination's DNS entry has been modified.

### PR DESCRIPTION
**Pull request by lorban**
Closes #5799

### CODE REVIEW

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 1]
```diff
@@ -90,6 +109,17 @@ protected void doStop() throws Exception
...
 
+    @ManagedAttribute(value = "The maximum duration in milliseconds a connection can be used for before it gets closed")
+    public long getMaxDuration()
+    {
+        return maxDuration;
+    }
+
+    public void setMaxDuration(long maxDuration)
+    {
+        this.maxDuration = maxDuration;
```
Comment (joakime) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: How does this get set?  I don't see a way of configuring this (or a testcase demonstrating how to set it)

**Review by joakime — COMMENTED** [review]
This reads a lot easier.
Not many changes to multiplex needed?

Some unit tests would be next step (if @sbordet likes this approach)

**Comment by lorban** [comment]
@joakime This PR can be considered a draft of what needs to be done, but it's not yet functional.

There's a new concept that the connection pool now needs to deal with: the fact that the pool can now contain a connection that is unusable because it expired. I'm busy writing some tests that expose this behavior, and they're miserably failing ATM.

**Comment by gregw** [comment]
@lorban I'd really like the Pool to not need to know about expiry and it should all be done in the ConnectionPool.
The ConnectionPool can check expiry whenever it acquires an entry and can also proactively sweep with acquire-check-release or check-acquire-recheck-release style

**Comment by lorban** [comment]
@gregw as discussed, this is what's going to happen. I've just closed the pool PR to reflect this.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 1 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: I've now added a new parameter to `ConnectionPoolTest` that is running the test suite against a pool with a configured max duration.

**Comment by lorban** [comment]
Side note: while working on this, I noticed thatthe low-level pool was missing one bug fix for  did not make it to the 9.4.x branch, and some closing behavior was not well tested. Maybe that should be moved to a different PR?

**Review by joakime — APPROVED** [review]
+1 from me

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 2]
```diff
@@ -48,8 +50,25 @@
 
     private final HttpDestination destination;
     private final Callback requester;
-    private final Pool<Connection> pool;
+    private final Pool<ConnectionHolder> pool;
     private boolean maximizeConnections;
+    private volatile long maxDurationNanos = 0L;
+
+    public static class ConnectionHolder
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Should be `protected`, no?

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 3]
```diff
@@ -281,16 +316,25 @@ protected void proceed()
...
         {
+            Pool<ConnectionHolder>.Entry entry = pool.acquire();
+            if (entry == null)
+                return null;
+            Connection connection = entry.getPooled().connection;
+
+            long maxDurationNanos = this.maxDurationNanos;
+            if (maxDurationNanos > 0L && entry.getPooled().isExpired(maxDurationNanos))
+            {
+                IO.close(connection);
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Please add a debug statement before this line to log it's been closed due to expiration.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 4]
```diff
@@ -426,12 +471,12 @@ public boolean sweep()
     {
         pool.values().stream().filter(entry -> entry.getPooled() instanceof Sweeper.Sweepable).forEach(entry ->
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: The `filter()` lambda is wrong, it's `entry.getPooled.connection` that may implement `Sweepable`.
Perhaps it's better to map to `connection` first, then filter, then `forEach()`

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 5]
```diff
@@ -71,6 +71,12 @@
     {
         return Stream.of(
             new ConnectionPoolFactory("duplex", destination -> new DuplexConnectionPool(destination, destination.getHttpClient().getMaxConnectionsPerDestination(), destination)),
+            new ConnectionPoolFactory("duplex-maxDuration", destination ->
+            {
+                DuplexConnectionPool pool = new DuplexConnectionPool(destination, destination.getHttpClient().getMaxConnectionsPerDestination(), destination);
+                pool.setMaxDuration(10);
+                return pool;
+            }),
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java:84]: I don't see a test that verifies that the max duration mechanism works?

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 6]
```diff
@@ -604,8 +604,8 @@ boolean tryRelease()
         }
 
         /**
-         * Try to mark the entry as removed.
-         * @return true if the entry has to be removed from the containing pool, false otherwise.
+         * Mark the entry as closed.
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: This change is confusing as says that it "mark" but actually "tries to mark", says "closed", but the method is called "tryRemove".
I would revert this change and perhaps explain a little more in the javadocs what the method actually does.

File: jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java [hunk 7]
```diff
@@ -53,7 +52,24 @@
 
     interface Factory
     {
-        Pool<String> getPool(int maxSize);
+        Pool<CloseableHolder> getPool(int maxSize);
```
Comment (sbordet) [inline jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:54]: Changes to this class are unrelated with the PR, right?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 2 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Indeed, let me change this.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 3 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Ok.

File: jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java (hunk 7 above)
Comment (lorban) [reply jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:54]: Correct. I wanted to add a few more checks to assert that connections are closed when they should be while working on this PR.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java (hunk 6 above)
Comment (lorban) [reply jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java]: Damn, I replaced incomplete doc with with even worse doc! Let me rework this.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 4 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Good catch, let me do that.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 5 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java:84]: You're right, adding a specific test for this functionality would be good. Let me add one.

**Review by sbordet — CHANGES_REQUESTED** [review]
@lorban as discussed, `ConnectionHolder` leaks as a type parameter to subclass constructors, which is not ideal.

Let's give a go at making `Pool.Entry` implement `Attachable` and store the duration inside that attachment, so the implementation is completely hidden in `AbstractConnectionPool`.

If  `maxDuration` is disabled, we don't even attach anything, and we save an allocation.

**Comment by lorban** [comment]
@sbordet Since `Connection` already is attachable, instead of attaching `Entry` instances, we can attach an `EntryHolder` instance instead and leave `Pool.Entry` untouched.

I went that route as I find it cleaner, even if that means an unnecessary extra allocation when expiration isn't configured.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 8]
```diff
@@ -54,6 +57,23 @@
     private final Callback requester;
     private final Pool<Connection> pool;
     private boolean maximizeConnections;
+    private volatile long maxDurationNanos = 0L;
+
+    private static class EntryHolder
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Can you please move this class at end of file?

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 9]
```diff
@@ -105,6 +125,22 @@ protected void doStop() throws Exception
         return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
     }
 
+    @ManagedAttribute(value = "The maximum duration in milliseconds a connection can be used for before it gets closed")
+    public long getMaxDuration()
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java:123]: Move the javadoc from the setter to the getter.
Also needs a paragraph explaining that the duration is passive and enforced on `acquire()`, so don't expect connections to be closed when `maxDuration` expires when they are idle -- there is the idleTimeout for that (link it).

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 10]
```diff
@@ -290,16 +326,35 @@ protected void proceed()
...
+            Pool<Connection>.Entry entry = pool.acquire();
+            if (entry != null)
+            {
+                Connection connection = entry.getPooled();
+
+                long maxDurationNanos = this.maxDurationNanos;
+                if (maxDurationNanos > 0L)
+                {
+                    EntryHolder holder = (EntryHolder)((Attachable)connection).getAttachment();
+                    if (holder != null && holder.isExpired(maxDurationNanos))
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Can `holder` ever be `null` here?

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 11]
```diff
@@ -432,6 +441,128 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
         assertThat(connectionPool.getConnectionCount(), Matchers.greaterThanOrEqualTo(count));
     }
 
+    @Test
+    public void testMaxDurationConnectionsWithConstrainedPool() throws Exception
+    {
+        // ConnectionPools may NOT open a few more connections than expected.
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: The comment is now wrong :)

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 12]
```diff
@@ -432,6 +441,128 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
...
+        transport.setConnectionPoolFactory(factory.factory);
+        client = new HttpClient(transport, null);
+        client.setExecutor(clientThreads);
+        client.start();
+
+        // Use the connection pool 5 times with a delay that is longer than the max duration in between each time.
+        for (int i = 0; i < 5; i++)
+        {
+            CountDownLatch latch = new CountDownLatch(1);
+            clientThreads.execute(() -> client.newRequest("localhost", connector.getLocalPort())
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Why `execute()` if you're waiting for the response? I would simplify and just:
```
ContentResponse response = client.newRequest(...)
  .timeout(5, SECONDS)
  .send();
assertEquals(200, response.getStatus());
```

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 13]
```diff
@@ -432,6 +441,128 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
...
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Don't `execute()`, see other comment.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+        {
+            client.stop();
+            client = null;
+        }
+    }
+
+    @Test
+    public void testMaxDurationConnectionsWithMultiplexedPool() throws Exception
+    {
+        // ConnectionPools may open a few more connections than expected.
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Remove comment, the comment below just before the asserts is enough.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 15]
```diff
@@ -432,6 +441,128 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
...
+                    poolRemoveCounter.incrementAndGet();
+                }
+            };
+            pool.setMaxDuration(maxDuration);
+            return pool;
+        });
+
+        startServer(new EmptyServerHandler());
+
+        QueuedThreadPool clientThreads = new QueuedThreadPool(3);
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Don't limit the threads to just so few. We don't want this test to depend on the thread pool behavior.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 16]
```diff
@@ -432,6 +441,128 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
...
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Same as above.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+                    poolRemoveCounter.incrementAndGet();
+                }
+            };
+            connectionPool.setMaxDuration(maxDuration);
+            return connectionPool;
+        });
+
+        startServer(new EmptyServerHandler());
+
+        QueuedThreadPool clientThreads = new QueuedThreadPool(3);
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Use the default maxThreads.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+
+            Thread.sleep(maxDuration * 2);
+        }
+
+        // Check that the pool acquired 5 and removed 4 connections;
+        // it must be exactly 4 removed b/c while the pool is not
+        // constrained, it can multiplex requests on a single connection
+        // so that should prevent opening more connections than needed.
+        assertThat(poolAcquireCounter.get(), is(5));
+        assertThat(poolRemoveCounter.get(), is(4));
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java:149]: I'm surprised the test has the same numbers of the non-multiplexed one.
If you're testing correct behavior, should you not test `connection` rather than number of times `acquired()` has been called (as it could be called with the same `connection` object)?

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+                }));
+            clientThreads.execute(() -> client.newRequest("http://localhost:" + connector.getLocalPort() + "/block")
+                .send(result ->
+                {
+                    if (result.isSucceeded())
+                    {
+                        latch1.countDown();
+                        latch2.countDown();
+                    }
+                }));
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Don't `execute()` also because you risk creating 2 connections, not just one.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 8 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Ok.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 9 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java:123]: Ok, I've moved the javadoc and made it more substantial.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java (hunk 10 above)
Comment (lorban) [reply jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: No, it can't. The only time the attachment can be null is before the entry is enabled or after it is removed so `pool.acquire()` can never return a connection with a null attachment. Let me remove this unneeded check.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 11 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: It's not because in this test, the pool is constrained to only a single connection. But I'll mention that in the comment.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 12 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Indeed. Let me change that.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 15 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Indeed. Let me change that.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 16 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Indeed. Let me change that.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 17 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Ok.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 18 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java:149]: This test should assert creations, not acquisitions. But besides that, it asserts that multiplexed connections are not reused since they expire between each use. This is why the figures are the same as for the non-multiplexed tests.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 19 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: I've completely revisited this test as there were many discrepancies between what I wanted to test and what it actually tested. It should now be much stricter about the behavior it generates and what it asserts.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 20]
```diff
@@ -432,6 +441,115 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
...
+        final int maxDuration = 30;
+        AtomicInteger poolAcquireCounter = new AtomicInteger();
+        AtomicInteger poolRemoveCounter = new AtomicInteger();
+        ConnectionPoolFactory factory = new ConnectionPoolFactory("duplex-maxDuration", destination ->
+        {
+            // Constrain the max pool size to 1.
+            DuplexConnectionPool pool = new DuplexConnectionPool(destination, 1, destination)
+            {
+                @Override
+                protected void acquired(Connection connection)
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: You want to override `onCreated()` not `acquired()`.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 21]
```diff
@@ -432,6 +441,115 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
...
+    {
+        // ConnectionPool may NOT open more connections than expected because
+        // it is constrained to a single connection in this test.
+
+        final int maxDuration = 30;
+        AtomicInteger poolAcquireCounter = new AtomicInteger();
+        AtomicInteger poolRemoveCounter = new AtomicInteger();
+        ConnectionPoolFactory factory = new ConnectionPoolFactory("duplex-maxDuration", destination ->
+        {
+            // Constrain the max pool size to 1.
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java:456]: Put it into a local variable, less "magic" numbers.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java [hunk 22]
```diff
@@ -432,6 +441,115 @@ protected void service(String target, org.eclipse.jetty.server.Request jettyRequ
...
+
+        final int maxDuration = 30;
+        AtomicInteger poolAcquireCounter = new AtomicInteger();
+        AtomicInteger poolRemoveCounter = new AtomicInteger();
+        ConnectionPoolFactory factory = new ConnectionPoolFactory("duplex-maxDuration", destination ->
+        {
+            DuplexConnectionPool pool = new DuplexConnectionPool(destination, destination.getHttpClient().getMaxConnectionsPerDestination(), destination)
+            {
+                @Override
+                protected void acquired(Connection connection)
```
Comment (sbordet) [inline jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: You want to override `onCreated()` not `acquired()`.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 23]
```diff
@@ -0,0 +1,306 @@
...
+        final int maxDuration = 30;
+        AtomicInteger poolCreateCounter = new AtomicInteger();
+        AtomicInteger poolRemoveCounter = new AtomicInteger();
+        AtomicReference<Pool<Connection>> poolRef = new AtomicReference<>();
+        ConnectionPoolFactory factory = new ConnectionPoolFactory("duplex-maxDuration", destination ->
+        {
+            int maxConnections = destination.getHttpClient().getMaxConnectionsPerDestination();
+            Pool<Connection> pool = new Pool<>(Pool.StrategyType.FIRST, maxConnections, false);
+            poolRef.set(pool);
+            MultiplexConnectionPool connectionPool = new MultiplexConnectionPool(destination, pool, destination, 2)
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Put `maxMultiplex` into a variable so it's clearer.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 24]
```diff
@@ -0,0 +1,306 @@
...
+
+        AtomicInteger poolCreateCounter = new AtomicInteger();
+        AtomicInteger poolRemoveCounter = new AtomicInteger();
+        AtomicReference<Pool<Connection>> poolRef = new AtomicReference<>();
+        ConnectionPoolFactory factory = new ConnectionPoolFactory("duplex-maxDuration", destination ->
+        {
+            int maxConnections = destination.getHttpClient().getMaxConnectionsPerDestination();
+            Pool<Connection> pool = new Pool<>(Pool.StrategyType.FIRST, maxConnections, false);
+            poolRef.set(pool);
+            MultiplexConnectionPool connectionPool = new MultiplexConnectionPool(destination, pool, destination, 2)
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Put `maxMultiplex` into a variable.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 25]
```diff
@@ -0,0 +1,306 @@
...
+        transport.setConnectionPoolFactory(factory.factory);
+        client = new HttpClient(transport, null);
+        client.setExecutor(clientThreads);
+        client.setIdleTimeout(maxIdle);
+        client.start();
+
+        CountDownLatch latch1 = new CountDownLatch(1);
+        CountDownLatch latch2 = new CountDownLatch(2);
+        // create 2 requests that are going to consume all the multiplexing slots
+        clientThreads.execute(() -> client.newRequest("http://localhost:" + connector.getLocalPort() + "/block")
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: No need to `execute()`.
Please use:
```
newRequest("localhost", port)
  .path("/block")
```

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+                {
+                    latch1.countDown();
+                    latch2.countDown();
+                }
+            }));
+
+        // wait for the 1st request to be serviced to make sure only 1 connection gets created
+        handlerSignalingSemaphore.acquire();
+
+        clientThreads.execute(() -> client.newRequest("http://localhost:" + connector.getLocalPort() + "/block")
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: No need to `execute()`.
Please use:
```
newRequest("localhost", port)
  .path("/block")
```

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java [hunk 27]
```diff
@@ -0,0 +1,306 @@
...
+        assertThat(poolRef.get().getInUseCount(), is(1));
+        assertThat(poolRef.get().getIdleCount(), is(0));
+        assertThat(poolRef.get().getClosedCount(), is(0));
+        assertThat(poolRef.get().size(), is(1));
+
+        // wait for the connection to expire
+        Thread.sleep(maxDuration + 500);
+
+        // send a 3rd request that will close the expired multiplexed connection
+        ContentResponse response = client.newRequest("localhost", connector.getLocalPort())
```
Comment (sbordet) [inline jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java:262]: Be explicit and add `.path("/dontblock")`.

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 28]
```diff
@@ -105,6 +109,27 @@ protected void doStop() throws Exception
         return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
     }
 
+    /**
+     * Get the max usage duration in milliseconds of the pool's connections.
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java]: Wrap with `<p>`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 20 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Indeed, let me change that.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 22 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java]: Indeed, let me change that.

File: jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java (hunk 21 above)
Comment (lorban) [reply jetty-client/src/test/java/org/eclipse/jetty/client/ConnectionPoolTest.java:456]: Ok.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 23 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Ok.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 24 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Ok.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 25 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Ok.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 26 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java]: Ok.

File: jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java (hunk 27 above)
Comment (lorban) [reply jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java:262]: Ok.

**Review by sbordet — APPROVED** [review]
